lib/polyfill/patchedmediakeys_safari.js

```diff
@@ -549,6 +526,21 @@ shaka.polyfill.PatchedMediaKeysSafari.MediaKeySession.prototype.
     onWebkitKeyMessage_ = function(event) {
   shaka.log.debug('PatchedMediaKeysSafari.onWebkitKeyMessage_', event);
 
+  // The first keymessage event from the Fairplay CDM is actually a request
+  // for the Fairplay certificate. We provide it to update() and wait for
+  // the next message, which will be the real license challenge.
+  var isCertificateRequest = shaka.util.Uint8ArrayUtils.equal(event.message,
+      new Uint8Array(shaka.util.StringUtils.toUTF8('certificate')));
```

Comment:
In a real app, where does the certificate come from?  Can the license (or another) server handle this message and give us the certificate?  This happens in the Chrome CDM where the first request may be a certificate request and the license server is expected to give it to us.  It would be nice to avoid a special case for both us and the app.

Reply:
I am not a Fairplay / license proxy expert so I asked our CWIP guy down the hall, and he tells me that it is not comparable to the Widevine flow. The Fairplay license proxy expects that the license challenge is already signed with the certificate.

Reply:
Is your concern that the app will now be required to provide a certificate to Shaka for Fairplay playback, whereas it is optional for Widevine? Is this what you mean by "special case"?

EME says:

>**setServerCertificate**
Provides a server certificate to be used to encrypt messages to the license server.
>
>Key Systems that use such certificates must also support requesting the certificate from the server via the Queue a "message" Event algorithm.

Since Safari doesn't implement the latest spec, and doesn't have this API, and we can't predict whether Apple will eventually implement it, or if the implementation will follow the spec, it may not be possible to avoid a special case for Safari.